d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,583 @@
+          });
+        }
+
+        @Override
+        public void onError(Throwable t)
+        {
+          _executorService.execute(() -> handleRpcError(t));
+        }
+
+        @Override
+        public void onCompleted()
+        {
+          _executorService.execute(() -> handleRpcCompleted());
+        }
+      };
+      _requestWriter = _stub.withWaitForReady().deltaAggregatedResources(responseReader);
+    }
+
+
+    /**
+     * Sends a client-initiated discovery request.
+     */
+    private void sendDiscoveryRequest(ResourceType type, Collection<String> resources)
+    {
+      _log.info("Sending {} request for resources: {}", type, resources);
+      DiscoveryRequestData request = new DiscoveryRequestData(_node, type, resources);
+      _requestWriter.onNext(request.toEnvoyProto());
+      _log.debug("Sent DiscoveryRequest\n{}", request);
+    }
+
+    private void sendAckRequest(ResourceType resourceType, String nonce)
+    {
+      AckOrNack ack = new AckOrNack(_node, resourceType, nonce);
+      _requestWriter.onNext(ack.toEnvoyProto());
+      _log.debug("Sent Ack\n{}", ack);
+    }
+
+    private void sendNackRequest(ResourceType resourceType, String nonce, @Nullable String errorDetail)
+    {
+      com.google.rpc.Status error = null;
+      if (errorDetail != null)
+      {
+        error = com.google.rpc.Status.newBuilder().setCode(Code.INVALID_ARGUMENT_VALUE).setMessage(errorDetail).build();
+      }
+      AckOrNack ack = new AckOrNack(_node, resourceType, nonce, error);
+      _requestWriter.onNext(ack.toEnvoyProto());
+      _log.debug("Sent Nack\n{}", ack);
+    }
+
+    private void handleResponse(DiscoveryResponseData response)
+    {
+      if (_closed)
+      {
+        return;
+      }
+      String respNonce = response.getNonce();
+      ResourceType resourceType = response.getResourceType();
+      switch (resourceType)
+      {
+        case D2_NODE:
+          handleD2NodeResponse(response);
+          break;
+        case D2_NODE_MAP:
+          handleD2NodeMapResponse(response);
+          break;
+        case UNKNOWN:
+          _log.warn("Received an unknown type of DiscoveryResponse\n{}", respNonce);
+          break;
+        default:
+          throw new AssertionError("Missing case in enum switch: " + resourceType);
+      }
+    }
+
+    private void handleRpcError(Throwable t)
+    {
+      handleRpcStreamClosed(Status.fromThrowable(t));
+    }
+
+    private void handleRpcCompleted()
+    {
+      handleRpcStreamClosed(Status.UNAVAILABLE.withDescription("ADS stream closed by server"));
+    }
+
+    private void handleRpcStreamClosed(Status error)
+    {
+      if (_closed)
+      {
+        return;
+      }
+      _log.error("ADS stream closed with status {}: {}. Cause: {}", error.getCode(), error.getDescription(),
+          error.getCause());
+      _closed = true;
+      handleStreamClosed(error);
+      cleanUp();
+      if (_responseReceived || _retryBackoffPolicy == null) {
```

Comment:
Any particular reason to gate exponential backoff by _responseReceived? Or, should we just always do exp-backoff and reset it upon successful connection?

Reply:
The responseReceived here is exactly trying to handle the backoff reset upon successful connection. The successful reconnection is defined as it has received a successful ADS response.